cmd/state.go

```diff
@@ -0,0 +1,167 @@
+	// interactively. If we're not, early exit.
+	if !cmdutil.Interactive() {
+		return nil, errors.Errorf("Resource URN %q ambiguously refers to multiple resources", urn)
+	}
+
+	surveycore.DisableColor = true
+	surveycore.QuestionIcon = ""
+	surveycore.SelectFocusIcon = opts.Color.Colorize(colors.BrightGreen + ">" + colors.Reset)
+	prompt := "Multiple resources with the given URN exist, please select the one to edit:"
+	prompt = opts.Color.Colorize(colors.SpecPrompt + prompt + colors.Reset)
+
+	var options []string
+	optionMap := make(map[string]*resource.State)
+	for _, ambiguousResource := range candidateResources {
+		// Prompt the user to select from a list of IDs, since these resources are known to all have the same URN.
+		message := fmt.Sprintf("Resource %q", ambiguousResource.ID)
+		if ambiguousResource.Protect {
+			message += " (Protected)"
+		}
+
+		if ambiguousResource.Delete {
+			message += " (Pending Deletion)"
+		}
+
+		options = append(options, message)
+		optionMap[message] = ambiguousResource
+	}
+
+	var option string
+	if err := survey.AskOne(&survey.Select{
+		Message:  prompt,
+		Options:  options,
+		PageSize: len(options),
+	}, &option, nil); err != nil {
+		return nil, errors.New("no resource selected")
+	}
+
+	return optionMap[option], nil
+}
+
+// runStateEdit runs the given state edit function on a resource with the given URN in the current stack.
+func runStateEdit(urn resource.URN, operation func(snap *deploy.Snapshot, res *resource.State) error) error {
+	return runTotalStateEdit(func(opts display.Options, snap *deploy.Snapshot) error {
+		res, err := locateStackResource(opts, snap, urn)
+		if err != nil {
+			return err
+		}
+
+		if res == nil {
+			return errors.Errorf("No such resource %q exists in the current state", urn)
+		}
+
+		return operation(snap, res)
+	})
+}
+
+// runTotalStateEdit runs a snapshot-mutating function on the entirity of the current stack's snapshot. Before mutating
+// the snapshot, the user is prompted for confirmation if the current session is interactive.
+func runTotalStateEdit(operation func(opts display.Options, snap *deploy.Snapshot) error) error {
+	opts := display.Options{
+		Color: cmdutil.GetGlobalColorization(),
+	}
+	s, err := requireCurrentStack(true, opts, true /*setCurrent*/)
+	if err != nil {
+		return err
+	}
+	snap, err := s.Snapshot(commandContext())
+	if err != nil {
+		return err
+	}
+
+	if cmdutil.Interactive() {
+		confirm := false
+		surveycore.DisableColor = true
+		surveycore.QuestionIcon = ""
+		surveycore.SelectFocusIcon = opts.Color.Colorize(colors.BrightGreen + ">" + colors.Reset)
+		if err = survey.AskOne(&survey.Confirm{
+			Message: "This command will edit your stack's state directly. Confirm?",
+		}, &confirm, nil); err != nil || !confirm {
+			return errors.New("confirmation declined")
+		}
+	}
+
+	stackIsAlreadyHosed := snap.VerifyIntegrity() != nil
```

Comment:
move down since you don't check this value till later.

Reply:
It's up here because `operation` mutates the snapshot in-place. The purpose of this check and the assert that follows is to make sure that, if a snapshot was valid, that `operation` doesn't change it in such a way that it's now invalid.

Reply:
ickkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkk